clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/MigrateSecurityGroupStrategy.java

```diff
@@ -0,0 +1,389 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.aws.deploy.handlers;
+
+import com.amazonaws.services.ec2.AmazonEC2;
+import com.amazonaws.services.ec2.model.IpPermission;
+import com.amazonaws.services.ec2.model.SecurityGroup;
+import com.amazonaws.services.ec2.model.Vpc;
+import com.netflix.spinnaker.clouddriver.aws.deploy.description.AbstractAmazonCredentialsDescription;
+import com.netflix.spinnaker.clouddriver.aws.deploy.description.UpsertSecurityGroupDescription;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.MigrateSecurityGroupReference;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.MigrateSecurityGroupResult;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.SecurityGroupIngressConverter;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.SecurityGroupLookupFactory.SecurityGroupLookup;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.SecurityGroupLookupFactory.SecurityGroupUpdater;
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.securitygroup.SecurityGroupMigrator.SecurityGroupLocation;
+import com.netflix.spinnaker.clouddriver.aws.provider.view.AmazonVpcProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.NetflixAmazonCredentials;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public abstract class MigrateSecurityGroupStrategy {
+
+  protected SecurityGroupLookup sourceLookup;
+  protected SecurityGroupLookup targetLookup;
+
+  abstract AmazonClientProvider getAmazonClientProvider();
+
+  /**
+   * Generates a result set, describing the actions required to migrate the source security group to the target.
+   *
+   * @param source                the source security group
+   * @param target                the target location
+   * @param sourceLookup          a lookup cache for the source region
+   * @param targetLookup          a lookup cache for the target region (may be the same as the sourceLookup)
+   * @param createIfSourceMissing indicates whether the operation should proceed if the source does not already exist;
+   *                              if false and the source group cannot be found, an IllegalStateException will be thrown
+   * @return the result set
+   */
+  public MigrateSecurityGroupResult generateResults(SecurityGroupLocation source, SecurityGroupLocation target,
```

Comment:
can the other public methods in this class (besides this one) become protected?  I think the only entry point should be via this method